File: G67 CP Quiz/G67 Grader 02/Stable Sort/a67_q2a_stable_sort.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class a67_q2a_stable_sort
{
    static int n,m;
    static int[] A;
    static StringBuilder result = new StringBuilder();
    public static void main(String[] args) throws IOException
    {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        
        StringTokenizer st = new StringTokenizer(reader.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(reader.readLine());
        A = new int[n];
        for(int i = 0; i < n; i++)
        {
            A[i] = Integer.parseInt(st.nextToken());
        }

        for(int q = 0; q < m; q++)
        {
            String[] data = reader.readLine().trim().split(" ");

            boolean isSorted = true;
            boolean isStable = true;
            for(int i = 1; i < n; i++)
            {
                int prevIdx = Integer.parseInt(data[i - 1]) - 1;
                int currentIdx = Integer.parseInt(data[i]) - 1;
                int prevValue = A[prevIdx];
                int currentValue = A[currentIdx];

                if(currentValue < prevValue)
                {
                    isSorted = false;
                    isStable = false;
                    break;
                }
                if(currentValue == prevValue && currentIdx < prevIdx)
                {
                    isStable = false;
                }
            }

            if(isSorted) { result.append("1 "); }
            else { result.append("0 "); }
            if(isStable) { result.append("1\n"); }
            else { result.append("0\n"); }
        }
        
        System.out.print(result);
    }
}

File: G67 CP Quiz/G67 Grader 02/Stable Sort/a67_q2a_stable_sort.py
import sys
n,m = map(int, sys.stdin.readline().split())
A = list(map(int, sys.stdin.readline().split()))

for _ in range(m):
    idx = [int(e) - 1 for e in sys.stdin.readline().split()]
    isSorted = True
    isStable = True
    
    for i in range(1, n):
        if(A[idx[i]] < A[idx[i - 1]]):
            isSorted = False
            isStable = False
            break
        if(A[idx[i]] == A[idx[i - 1]] and idx[i] < idx[i - 1]):
            isStable = False
    
    if(isSorted):
        sys.stdout.write("1 ")
    else:
        sys.stdout.write("0 ")
    if(isStable):
        sys.stdout.write("1\n")
    else:
        sys.stdout.write("0\n")
